source/alloc/src/heap.rs

```diff
@@ -1,505 +1,128 @@
 use linked_list_allocator::Heap;
-use maitake::sync::WaitQueue;
-
-/// An Anachro Heap item
-pub struct AHeap {
-    freelist: MpscQueue<Recycle>,
-    state: AtomicU8,
-    heap: UnsafeCell<Heap>,
-    heap_wait: WaitQueue,
-    inhibit_alloc: AtomicBool,
-    any_frees: AtomicBool,
-}
-
-// SAFETY: Safety is checked through the `state` member, which uses
-// atomic operations to ensure the data is initialized and exclusively
-// accessed.
-unsafe impl Sync for AHeap {}
-unsafe impl Send for AHeap {}
-
-impl AHeap {
-    /// The AHeap is initialized, and no `HeapGuard`s are active.
-    const INIT_IDLE: u8 = 1;
-
-    /// The AHeap is "locked", and cannot currently be retrieved. In MOST cases
-    /// this also means the heap is initialized, except for the brief period of
-    /// time while the heap is being initialized.
-    const BUSY_LOCKED: u8 = 2;
+use maitake::sync::{Mutex, WaitQueue};
 
-    /// Construct a thread safe async allocator from a pool of memory.
-    ///
-    /// Safety: The pool of memory MUST be valid for the 'static lifetime, e.g.
-    /// obtained by a leaked buffer, a linker section, or some other mechanism.
-    /// Additionally, we must semantically have exclusive access to this region
-    /// of memory: There must be no other live references, or pointers to this
-    /// region that are dereferenced after this call.
-    pub unsafe fn bootstrap(addr: *mut u8, size: usize) -> Result<(NonNull<Self>, HeapGuard), ()> {
-        // First, we go all bump-allocator to emplace ourselves within this region
-        let mut cursor = addr;
-        let end = (addr as usize).checked_add(size).ok_or(())?;
-        let mut used = 0;
-
-        let stub_ptr;
-        let aheap_ptr;
-
-        // We start with the stub node required for our mpsc queue.
-        {
-            let stub_layout = Layout::new::<Recycle>();
-            let stub_offset = cursor.align_offset(stub_layout.align());
-            let stub_size = stub_layout.size();
-            used += stub_offset;
-            used += stub_size;
-
-            if used > size {
-                return Err(());
-            }
-
-            cursor = cursor.wrapping_add(stub_offset);
-            stub_ptr = cursor.cast::<Recycle>();
-            stub_ptr.write(Recycle {
-                links: Links::new_stub(),
-                node_layout: stub_layout,
-            });
-            cursor = cursor.add(stub_size);
-        }
+static OOM_WAITER: WaitQueue = WaitQueue::new();
+static INHIBIT_ALLOC: AtomicBool = AtomicBool::new(false);
 
-        // Next we allocate ourselves
-        {
-            let aheap_layout = Layout::new::<Self>();
-            let aheap_offset = cursor.align_offset(aheap_layout.align());
-            let aheap_size = aheap_layout.size();
-            used += aheap_offset;
-            used += aheap_size;
-
-            if used > size {
-                return Err(());
+pub async fn alloc(layout: Layout) -> NonNull<u8> {
+    loop {
+        unsafe {
```

Comment:
very minor nit, not important: does this entire thing actually need to be inside the unsafe block? it's really just the call to `alloc::alloc::alloc` that's unsafe...